lib/hterm.js

```diff
@@ -388,6 +345,157 @@ hterm.Terminal.prototype.focusHyperCaret = function () {
   this.hyperCaret.focus();
 };
 
+// patch for unicode encapsulation
+// see comments in original code
+hterm.Screen.prototype.insertString = function (str) {
+  let cursorNode = this.cursorNode_;
+  let cursorNodeText = cursorNode.textContent;
+  this.cursorRowNode_.removeAttribute('line-overflow');
+  const strWidth = lib.wc.strWidth(str);
+  this.cursorPosition.column += strWidth;
+  let offset = this.cursorOffset_;
+  let reverseOffset = hterm.TextAttributes.nodeWidth(cursorNode) - offset;
+  if (reverseOffset < 0) {
+    const ws = lib.f.getWhitespace(-reverseOffset);
+    if (!(this.textAttributes.underline ||
+          this.textAttributes.strikethrough ||
+          this.textAttributes.background ||
+          this.textAttributes.wcNode ||
+          this.textAttributes.tileData !== null)) {
+      str = ws + str;
+    } else if (cursorNode.nodeType === 3 ||
+               !(cursorNode.wcNode ||
+                 cursorNode.tileNode ||
+                 cursorNode.style.textDecoration ||
+                 cursorNode.style.backgroundColor)) {
+      cursorNode.textContent = (cursorNodeText += ws);
+    } else {
+      const wsNode = cursorNode.ownerDocument.createTextNode(ws);
+      this.cursorRowNode_.insertBefore(wsNode, cursorNode.nextSibling);
+      this.cursorNode_ = cursorNode = wsNode;
+      this.cursorOffset_ = offset = -reverseOffset;
+      cursorNodeText = ws;
+    }
+    reverseOffset = 0;
+  }
+  this.wrapUnicode(cursorNode);
+  if (this.textAttributes.matchesContainer(cursorNode)) {
+    if (reverseOffset === 0) {
+      cursorNode.textContent = cursorNodeText + str;
+    } else if (offset === 0) {
+      cursorNode.textContent = str + cursorNodeText;
+    } else {
+      cursorNode.textContent =
+          hterm.TextAttributes.nodeSubstr(cursorNode, 0, offset) +
+          str + hterm.TextAttributes.nodeSubstr(cursorNode, offset);
+    }
+    this.cursorOffset_ += strWidth;
+    this.wrapUnicode(cursorNode);
+    return;
+  }
+  if (offset === 0) {
+    const previousSibling = cursorNode.previousSibling;
+    if (previousSibling &&
+        this.textAttributes.matchesContainer(previousSibling)) {
+      previousSibling.textContent += str;
+      this.cursorNode_ = previousSibling;
+      this.cursorOffset_ = lib.wc.strWidth(previousSibling.textContent);
+      this.wrapUnicode(previousSibling);
+      return;
+    }
+    const newNode = this.textAttributes.createContainer(str);
+    this.cursorRowNode_.insertBefore(newNode, cursorNode);
+    this.cursorNode_ = newNode;
+    this.cursorOffset_ = strWidth;
+    this.wrapUnicode(newNode);
+    return;
+  }
+  if (reverseOffset === 0) {
+    const nextSibling = cursorNode.nextSibling;
```

Comment:
I'd rather keep it like the original `hterm` code to make diffs easier. Makes sense?